crates/ruff_linter/src/rules/ruff/rules/unused_variable_accessed.rs

```diff
@@ -0,0 +1,124 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_semantic::Binding;
+use ruff_python_stdlib::builtins::is_python_builtin;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for usages of variables marked as unused (variable names starting with an underscore, except '_') in functions.
+/// Forbid using method and function variables that are marked as unused.
+///
+/// ## Why is this bad?
+/// Marking variables with a leading underscore conveys that they are intentionally unused within the function or method.
+/// When these variables are later referenced in the code, it causes confusion and potential misunderstandings about
+/// the code's intention. A variable marked as "unused" but subsequently used suggests oversight or unintentional use
+/// and detracts from the clarity and maintainability of the codebase.
+///
+/// ## Example
+/// ```python
+/// def function():
+///     _variable = 3
+///     return _variable + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def function():
+///     variable = 3
+///     return variable + 1
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct UnusedVariableAccessed {
+    name: String,
+    // fix: String,
+    shadowed_kind: ShadowedKind,
+}
+
+impl Violation for UnusedVariableAccessed {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Local variable `{}` marked as unused is accessed",
+            self.name
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(match self.shadowed_kind {
+            ShadowedKind::BuiltIn => {
+                "Consider using preferred trailing underscores to avoid shadowing a built-in."
+                    .to_string()
+            }
+            ShadowedKind::Some => {
+                "Consider using preferred trailing underscores to avoid shadowing a variable."
+                    .to_string()
+            }
+            ShadowedKind::None => "Consider removing leading underscores.".to_string(),
+        })
+    }
+}
+
+/// RUF052
+pub(crate) fn unused_variable_accessed(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    let name = binding.name(checker.source());
+
+    // only used variables
+    if !name.starts_with('_')
+        || name == "_"
+        || binding.is_unused()
+        || binding.is_global()
+        || binding.is_nonlocal()
+        || (!binding.kind.is_argument() && !binding.kind.is_assignment())
+        || !checker.semantic().scopes[binding.scope].kind.is_function()
+        || checker.settings.dummy_variable_rgx.is_match(name)
+    {
+        return None;
+    }
```

Comment:
There's quite a lot packed into this one condition here. I'd do it in a slightly more verbose way: I would find it more readable, as it allows you to add comments explaining the meaning of each check:
```suggestion
    // only variables marked as private
    if !name.starts_with('_') || name == "_" || is_dunder(name) {
        return None;
    }

    // only used variables
    if binding.is_unused() {
        return None;
    }

    // Only variables defined via function arguments or assignments.
    // This excludes `global` and `nonlocal` variables.
    if !matches!(
        binding.kind,
        BindingKind::Argument | BindingKind::Assignment
    ) {
        return None;
    }

    // Only variables defined in function scopes
    if !checker.semantic().scopes[binding.scope].kind.is_function() {
        return None;
    }

    if checker.settings.dummy_variable_rgx.is_match(name) {
        return None;
    }
```

Reply:
>The binding is `global.` This could be a binding created by a `global` statement or a
binding that shadows such a binding.
>For example, both of the bindings in the following function are `global`:
```python
 def f():
     global x
     x = 1
 ```
 
 I believe that it checks against new binding when assigned to a variable declared as global, and thus, it must be checked for its 'globalness' - 'nonlocalness'.

Reply:
yep, great catch! The semantic model is pretty subtle here; sorry for my mistake :-)